xrmocap/ops/bottom_up_association/matching/fourdag_matching.py

```diff
@@ -0,0 +1,770 @@
+                            self.m_boneNodes[paf_id][view].append((joint1_candidate, joint2_candidate)) 
+
+    def cal_bone_epi_edges(self):
+        for paf_id in range(self.n_pafs):
+            joint_pair = [self.paf_dict[0][paf_id], self.paf_dict[1][paf_id]]
+            for view1 in range(5-1):
+                for view2 in range(view1+1, 5):
+                    nodes1 = self.m_boneNodes[paf_id][view1]
+                    nodes2 = self.m_boneNodes[paf_id][view2]
+                    epi = np.full((len(nodes1), len(nodes2)), -1.0)
+                    for bone1_id in range(len(nodes1)):
+                        for bone2_id in range(len(nodes2)):
+                            node1 = nodes1[bone1_id]
+                            node2 = nodes2[bone2_id]
+                            epidist = np.zeros(2)
+                            for i in range(2):
+                                epidist[i] = self.m_epiEdges[joint_pair[i]][view1][view2][node1[i], node2[i]]
+                            if epidist.min() < 0:
+                                continue
+                            epi[bone1_id, bone2_id] = epidist.mean()
+                    self.m_boneEpiEdges[paf_id][view1][view2] = epi
+                    self.m_boneEpiEdges[paf_id][view2][view1] = epi.T
+
+    def cal_bone_temp_edges(self):
+        for paf_id in range(self.n_pafs):
+            joint_pair = [self.paf_dict[0][paf_id], self.paf_dict[1][paf_id]]
+            for view in range(self.n_views):
+                nodes = self.m_boneNodes[paf_id][view]
+                temp = np.full((len(self.last_multi_kps3d), len(nodes)), -1.0)
+                for pid in range(len(temp)):
+                    for node_candidate in range(len(nodes)):
+                        node = nodes[node_candidate]
+                        tempdist = []
+                        for i in range(2):
+                            tempdist.append(self.m_tempEdges[joint_pair[i]][view][pid][node[i]])
+                        if min(tempdist) > 0:
+                            temp[pid, node_candidate] = sum(tempdist) / len(tempdist)
+                self.m_boneTempEdges[paf_id][view] = temp
+
+    def initialize(self):
+        for joint_id in range(self.n_joints):
+            for view in range(self.n_views):
+                self.m_assignMap[view][joint_id] = np.full(len(self.kps2d_paf[view]['joints'][joint_id]), -1)
+        
+        self.m_personsMap = {}
+        for pid in range(len(self.last_multi_kps3d)):
+            self.m_personsMap[pid] = np.full((self.n_joints, self.n_views), -1)
+
+    def enum_cliques(self):
+        tmpCliques = {i:[] for i in range(self.n_pafs)}
+        for paf_id in range(self.n_pafs):
+            nodes = self.m_boneNodes[paf_id]
+            pick = [-1]* (self.n_views+1)
+            available_node = {i:{j:[] for j in range(self.n_views+1)} for i in range(self.n_views+1)}
```

Comment:
Does `self.n_views+1` mean including time domain information? If so, we could add a commet here.

Reply:
it means the 3d keypoints of the previous frame